### PR TITLE
Test utilities

### PR DESCRIPTION
**Pull request by CharliePoole**
These are changes I made to logging and test utilities in order to debug issue #2454. There are no functional changes as far as users are concerned.

I think it's all an improvement and some of it helped me figure out a solution to the above issue. I'm making it a separate PR so that the main one for #2454 can get merged quickly. However, it would be nice to get this one in as I'm building on it locally. 😄

### CODE REVIEW

File: src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs [hunk 1]
```diff
@@ -242,6 +242,11 @@ private void RestoreQueues()
 
         private void OnEndOfShift(object sender, EventArgs ea)
         {
+            if (StartNextShift())
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs]: This is a non-trivial change, should it be in this PR?

File: src/NUnitFramework/FakeTestMethod.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
+﻿// ***********************************************************************
+// Copyright (c) 2014 Charlie Poole, Rob Prouse
```
Comment (ChrisMaddock) [inline src/NUnitFramework/FakeTestMethod.cs]: Trivial - but wrong date here. 🙂

**Comment by CharliePoole** [comment]
@ChrisMaddock Your comment is understandable as I separated the changes from the things that motivated it, which were:

1. Issue #2454 - these are all changes I created in working on that issue in order to debug it. The fix is in a separate PR, of course.

2. Unit testing of parallelism - part of but going beyond #2454, I wanted to be able to run tests that examine how parallelism works without creating a separate assembly. I have not included any of those tests yet as it's a work in progress.

**Comment by CharliePoole** [comment]
On the placement of the classes: TestBuilder and the Fake methods are in the TestUtilities namespace, of course, but they are linked from the common parent directory NUnitFramework because they are used in both the framework and the nunitlite tests. Our past practice/convention has been that anything in that directory is shared between projects - for example the CommonAssemlblyInfo file. What you suggest is, of course, an alternative convention that we haven't been using lately but did use in the past.

File: src/NUnitFramework/FakeTestMethod.cs (hunk 2 above)
Comment (CharliePoole) [reply src/NUnitFramework/FakeTestMethod.cs]: That's the year the code was created in the old `Fakes.cs` file. Splitting it into a separate file without change didn't seem like a reason to update the year.

File: src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs (hunk 1 above)
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs]: I didn't intend it to be here - messed up separating my changes.

**Comment by ChrisMaddock** [comment]
Apologies - I was aware of the issues this PR related to, what I wanted to understand was the reason behind this refactor. e.g. `Fakes.GetTestMethod()` vs `new FakeTestMethod()`. I'm sure there is a good reason - I just wasn't sure what it was from just looking at the code! 😄 

> Our past practice/convention has been that anything in that directory is shared between projects - for example the CommonAssemlblyInfo file. What you suggest is, of course, an alternative convention that we haven't been using lately but did use in the past.

That would make more sense to me - but I don't have significantly strong opinion to push on it! 🙂

**Comment by CharliePoole** [comment]
It's definitely a change we could make (in file locations). Doing it as we do has the advantage that the path from the project directory is simply `..`

A possibility is to put a `TestUtilities` directory under `NUnitFramework`.

Regarding the motivation for changing, I think I'll add one sample test that may illustrate why I liked the use of `new` better than static constructors.

**Comment by CharliePoole** [comment]
I'm closing this without merging. The remaining changes will go into actual fixes of parallelism, rather than in a separate PR for the utilities.

**Pull request closed without merging**